spec/requests/api/v2/providers/courses/build_new_spec.rb

```diff
@@ -0,0 +1,139 @@
+      end
+    end
+
+    context 'with an invalid attribute set in query parameters' do
+      let(:params) do
+        { course: {
+          name: 'Foo Bar Course',
+          is_send: 'wibble', # invalid
+          maths: 'must_have_qualification_at_application_time',
+          english: 'must_have_qualification_at_application_time',
+        } }
+      end
+
+      it 'returns edit_options and errors but no model' do
+        response = do_get params
+        expect(response).to have_http_status(:unprocessable_entity)
+        jsonapi_response = parse_response(response)
+
+        expect(jsonapi_response['errors']).to eq [{
+                                            "detail" => "Is send is not included in the list",
+                                            "source" => {}, # todo: make this work?
+                                            "title" => "Invalid is_send"
+                                          }]
+        expect(jsonapi_response['meta']).not_to be_nil
+        expect(jsonapi_response['meta']['edit_options']).not_to be_nil
+        expect(jsonapi_response['meta']['edit_options']['study_modes']).to eq %w[full_time part_time full_time_or_part_time]
+      end
+    end
+
+    context 'with an invalid enum attribute set in query parameters' do
+      let(:params) do
+        { course: {
+          name: 'Foo Bar Course',
+          maths: 'wibble', # invalid
+          english: 'must_have_qualification_at_application_time',
+        } }
+      end
+
+      it 'returns edit_options and errors but no model' do
+        response = do_get params
+        expect(response).to have_http_status(:unprocessable_entity)
+        jsonapi_response = parse_response(response)
+
+        expect(jsonapi_response['errors']).to eq [{
+                                            "detail" => "Maths is invalid",
+                                            "source" => {},
+                                            "title" => "Invalid maths"
+                                          }]
+        expect(jsonapi_response['meta']['edit_options']['study_modes']).to eq %w[full_time part_time full_time_or_part_time]
+      end
+    end
+  end
+
+  def do_get(params)
+    get "/api/v2/recruitment_cycles/#{recruitment_cycle.year}" \
+          "/providers/#{provider.provider_code}" \
+          "/courses/build_new",
+        headers: { 'HTTP_AUTHORIZATION' => credentials },
+        params: params
+    response
+  end
+
+  def parse_response(response)
+    JSON.parse(response.body)
+  end
+
+  def course_to_jsonapi(course)
+    rendered_course = jsonapi_renderer.render(
+      course,
+      class: {
+        Course: API::V2::SerializableCourse
+      }
+    )
+    # Converting the rendered course to JSON and back normalises any symbols
+    # into strings. Better #deep_symbolize_keys because it also does values.
+    JSON.parse(rendered_course.to_json)
```

Comment:
maybe 
https://api.rubyonrails.org/v4.2/classes/ActiveSupport/HashWithIndifferentAccess.html

json parsing is normally expensive...

Reply:
This is copy pasted from other tests, I don't really understand it